crates/editor/src/element.rs

```diff
@@ -1558,6 +1589,258 @@ impl EditorElement {
+                )
+            } else {
+                boxes
+            };
+
+            let mut element = h_flex()
+                .id(SharedString::from(format!("diagnostic-{}", line_ix.0)))
+                .h(line_height)
+                .w_full()
+                .bg(sev_to_color(&elem.severity).color(cx).opacity(0.07))
+                .text_color(sev_to_color(&elem.severity).color(cx))
+                .text_sm()
+                .font_family(style.text.font().family)
+                .child(
+                    h_flex()
+                        .h(line_height)
+                        .gap_2()
+                        .child(boxes)
+                        .child(elem.message),
+                )
+                .into_any();
+
+            element.prepaint_as_root(point(elem.x, elem.y), AvailableSpace::min_size(), cx);
+            elements.insert(line_ix, element);
+        }
+
+        elements
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn gather_inline_diagnostics(
+        &self,
+        snapshot: &EditorSnapshot,
+        hitbox: &Hitbox,
+        line_layouts: &[LineWithInvisibles],
+        crease_trailers: &[Option<CreaseTrailerLayout>],
+        content_origin: gpui::Point<Pixels>,
+        scroll_pixel_position: gpui::Point<Pixels>,
+        line_height: Pixels,
+        em_width: Pixels,
+        start_row: DisplayRow,
+        end_row: DisplayRow,
+        cx: &mut WindowContext,
+    ) -> HashMap<DisplayRow, DiagnosticElement> {
+        let mut inline_diagnostics: HashMap<DisplayRow, DiagnosticElement> = HashMap::default();
+        let buffer_snapshot = snapshot.display_snapshot.buffer_snapshot.clone();
+
+        let start_point =
+            snapshot.display_point_to_point(DisplayPoint::new(start_row, 0), Bias::Left);
+        let end_point = snapshot.display_point_to_point(DisplayPoint::new(end_row, 0), Bias::Right);
+        let diagnostics = buffer_snapshot
+            .diagnostics_in_range(start_point..end_point, false)
+            .map(|DiagnosticEntry { diagnostic, range }| DiagnosticEntry {
+                diagnostic,
+                range: range.to_point(&buffer_snapshot),
+            })
+            .sorted_by_key(|diagnostic| {
+                (
+                    diagnostic.diagnostic.severity,
+                    std::cmp::Reverse(diagnostic.diagnostic.is_primary),
+                    diagnostic.range.start.row,
+                    diagnostic.range.start.column,
+                )
+            });
+
+        let max_entries = 5; // ToDo: Configuration
+        for diagnostic in diagnostics {
+            let diag_point = MultiBufferPoint::new(diagnostic.range.start.row, 0);
+            let display_point = snapshot.point_to_display_point(diag_point, Bias::Right);
+
+            if !(start_row..end_row).contains(&display_point.row()) {
+                continue;
+            }
+            let line_ix = DisplayRow(display_point.row().minus(start_row));
```

Comment:
NOTE: I'm not entirely sure why this works, as `display_point` seems to be converted from a diagnostics range start, hence should be ok without any subtractions.

Overall, we're safer with multi buffer anchors user as coordinates, esp. if we start to store date between frames.

Reply:
You and me both! All of this logic is basically the result of me squinting cross eyed at the various similar places I found and comparing type signatures of function arguments to get a conversion chain that appears to work. I have no idea how costly these calls are other than to say I've not been able to notice any lagging of the editor regardless of file size or diagnostic count (zero actual measurements here, I just haven't noticed anything).

That said, I've also not tested this thoroughly to see if loading up a buffer with tens of thousands of diagnostics would slow things down considerably. The only way I can think that it would is if there were many thousands of diagnostics that were somehow all visible in a single on screen view of the editor.